ta/supp_plugin/ta_entry.c

```diff
@@ -0,0 +1,180 @@
+	struct add_op {
+		uint32_t a;
+		uint32_t b;
+	} values = { 0 };
+
+	if (exp_pt != param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	values.a = params[0].value.a;
+	values.b = params[0].value.b;
+
+	res = tee_invoke_supp_plugin(&uuid, TEST_PLUGIN_PASS_VALUES,
+				     params[1].value.a, (void *)&values,
+				     sizeof(struct add_op), &outlen);
+	params[0].value.a = values.a;
+
+	return res;
+}
+
+static TEE_Result write_array(uint32_t param_types,
+			      TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+					  TEE_PARAM_TYPE_MEMREF_OUTPUT,
+					  TEE_PARAM_TYPE_NONE,
+					  TEE_PARAM_TYPE_NONE);
+	TEE_Result res = TEE_ERROR_GENERIC;
+	size_t outlen = 0;
+	char *buf = NULL;
+
+	if (exp_pt != param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	buf = malloc(params[0].memref.size);
+	if (!buf)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	/*
+	 * Plugins use a same buffer for input and output data.
+	 * Make copy of input data to avoid erasing of it by the output.
+	 *
+	 * Output data contain file name the input data will be stored.
+	 */
+	memcpy(buf, params[0].memref.buffer, params[0].memref.size);
+	res = tee_invoke_supp_plugin(&uuid, TEST_PLUGIN_WRITE_ARR, 0,
+				     buf, params[0].memref.size, &outlen);
+	memcpy(params[1].memref.buffer, buf, outlen);
+	free(buf);
+
+	return res;
+}
+
+static TEE_Result get_array(uint32_t param_types,
+			    TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_OUTPUT,
+					  TEE_PARAM_TYPE_VALUE_OUTPUT,
+					  TEE_PARAM_TYPE_NONE,
+					  TEE_PARAM_TYPE_NONE);
+	TEE_Result res = TEE_ERROR_GENERIC;
+	size_t outlen = 0;
+
+	if (exp_pt != param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = tee_invoke_supp_plugin(&uuid, TEST_PLUGIN_GET_ARR, 0,
+				     params[0].memref.buffer,
+				     params[0].memref.size,
+				     &outlen);
+	params[1].value.a = (uint32_t)outlen;
+
+	return res;
+}
+
+static TEE_Result bad_input_uuid(uint32_t param_types __unused,
+				 TEE_Param params[TEE_NUM_PARAMS] __unused)
+{
+	char data[32] = {  };
+	size_t inlen = sizeof(data);
+
+	return tee_invoke_supp_plugin(NULL, 0, 0, data, inlen, NULL);
+}
+
+static TEE_Result bad_input_data(uint32_t param_types __unused,
+				 TEE_Param params[TEE_NUM_PARAMS] __unused)
+{
+	return tee_invoke_supp_plugin(&uuid, 0, 0, NULL, 10, NULL);
+}
+
+static TEE_Result bad_input_inlen(uint32_t param_types __unused,
+				  TEE_Param params[TEE_NUM_PARAMS] __unused)
+{
+	char data[32] = {  };
+
+	return tee_invoke_supp_plugin(&uuid, 0, 0, data, 0, NULL);
+}
+
+static TEE_Result call_unknown_plugin(uint32_t param_types __unused,
+				      TEE_Param params[TEE_NUM_PARAMS]
+				      __unused)
+{
+	const TEE_UUID nulluuid = {  }; /* null-uuid */
+	char data[32] = {  };
+	size_t inlen = sizeof(data);
+
+	return tee_invoke_supp_plugin(&nulluuid, 0, 0, data, inlen, NULL);
+}
+
+TEE_Result TA_InvokeCommandEntryPoint(void *pSessionContext __unused,
```

Comment:
All these TA functions could be replaced by a single TA function (the os_test TA would be a good match then) with a flexible interface allowing to put most of the logic in xtest. If you'd rather keep it as is that's fine too.

Reply:
I'd prefer to keep it as it is, thanks